File: ideas/gpt-keyword/OpenAI/FunctionCallContext.cs
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//

using Azure.AI.OpenAI;
using System;
using System.Collections.Generic;

public class FunctionCallContext
{
    private FunctionFactory _functionFactory;
    private IList<ChatRequestMessage> _messages;
    private string _functionName = "";
    private string _functionArguments = "";

    public FunctionCallContext(FunctionFactory functionFactory, IList<ChatRequestMessage> messages)
    {
        _functionFactory = functionFactory;
        _messages = messages;
    }

    
    public bool CheckForUpdate(StreamingChatCompletionsUpdate update)
    {
        var updated = false;

        var name = update?.FunctionName;
        if (name != null)
        {
            _functionName = name;
            updated = true;
        }

        var args = update?.FunctionArgumentsUpdate;
        if (args != null)
        {
            _functionArguments += args;
            updated = true;
        }

        return updated;
    }

    public string? TryCallFunction()
    {
        var ok = _functionFactory.TryCallFunction(_functionName, _functionArguments, out var result);
        if (!ok) return null;

        Console.WriteLine($"\rassistant-function: {_functionName}({_functionArguments}) => {result}");
        Console.Write("\nAssistant: ");

        _messages.Add(new ChatRequestAssistantMessage("") { FunctionCall = new FunctionCall(_functionName, _functionArguments) });
        _messages.Add(new ChatRequestFunctionMessage(_functionName, result));

        return result;
    }

    public void Clear()
    {
        _functionName = "";
        _functionArguments = "";
    }
}

File: src/ai/.x/templates/openai-chat-webpage-with-speech-and-functions-ts/src/FunctionCallContext.ts
import { FunctionFactory } from "./FunctionFactory";
import { ChatCompletionMessageParam } from "openai/resources/index.js";
import { ChatCompletionChunk } from "openai/resources/index.js";

export class FunctionCallContext {
  private function_factory: FunctionFactory;
  private messages: Array<ChatCompletionMessageParam>;
  private function_name: string;
  private function_arguments: string;

  constructor(function_factory: FunctionFactory, messages: Array<ChatCompletionMessageParam>) {
    this.function_factory = function_factory;
    this.messages = messages;
    this.function_name = "";
    this.function_arguments = "";
  }

  checkForUpdate(choice: ChatCompletionChunk.Choice): boolean {
    let updated = false;

    const name = choice.delta?.function_call?.name;
    if (name !== undefined) {
      this.function_name = name;
      updated = true;
    }

    const args = choice.delta?.function_call?.arguments;
    if (args !== undefined) {
      this.function_arguments = `${this.function_arguments}${args}`;
      updated = true;
    }

    return updated;
  }

  tryCallFunction(): string | undefined {
    let result = this.function_factory.tryCallFunction(this.function_name, this.function_arguments);
    if (result === undefined) {
      return undefined;
    }

    console.log(`assistant-function: ${this.function_name}(${this.function_arguments}) => ${result}`);

    this.messages.push({ role: 'assistant', content: '', function_call: { name: this.function_name, arguments: this.function_arguments } });
    this.messages.push({ role: 'function', content: result, name: this.function_name });

    return result;
  }

  clear(): void {
    this.function_name = "";
    this.function_arguments = "";
  }
}